parachain/pallets/erc20-app/src/lib.rs

```diff
@@ -167,5 +188,42 @@ pub mod pallet {
 
 			Ok(())
 		}
+
+		#[pallet::weight(100_000_000)]
+		#[transactional]
+		pub fn create(origin: OriginFor<T>, token: H160) -> DispatchResult {
+			let who = T::CallOrigin::ensure_origin(origin)?;
+			if who != <Address<T>>::get() {
+				return Err(DispatchError::BadOrigin.into());
+			}
+
+			let asset_id = Self::make_asset_id(token);
+			T::Assets::create(asset_id, T::PalletId::get().into_account(), true, 1)?;
+
+			<AssetId<T>>::insert(token, asset_id);
+
+			Ok(())
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		fn make_asset_id(address: H160) -> u128 {
```

Comment:
@musnit @alistair-singh Just want to raise a specific design implication here for discussion:

The new assets pallet uses a 128-bit number for asset identifiers. So for each new wrapped asset, we have to generate a unique 128-bit ID. The options for this generation are to either keep an incrementing counter in storage, or to generate random numbers with a very low probability of collision. I'm opting for the latter as its stateless.

So the goal of this function to generate a random 128-bit number using various sources of entropy (The contract address, last K block hashes, etc).

The probability of collision is exactly the same as the scenario where random V4 UUIDs (also 128 bits) are used as database keys. According to this [wikipage](https://en.wikipedia.org/wiki/Universally_unique_identifier#Collisions):

> For example, the number of random version-4 UUIDs which need to be generated in order to have a 50% probability of at least one collision is 2.71 quintillion. This number is equivalent to generating 1 billion UUIDs per second for about 85 years. A file containing this many UUIDs, at 16 bytes per UUID, would be about 45 exabytes.

And

> The probability to find a duplicate within 103 trillion version-4 UUIDs is one in a billion.

Given that we'll only expect to see maybe a few dozen new assets per week, our probability of collision would be several magnitudes less than the above. Practically negligible. Like we'd need to operate the bridge for hundreds or thousands of years before seeing a collision.

Currently the worst case possibility is that the transferred funds would be lost on collision.

Reply:
We could store the original H160 as asset metadata since we are using assets pallet. We could then look up and use it for collision detection so that we can fail gracefully. Clients can retry the transaction and it would successful if we have different entropy.

Reply:
Thanks Alistair, that could work. Though I've now moved to simpler mechanism with less potential for negative scenarios. I've introduced a new pallet called `AssetRegistry`, which contains a counter in storage. Every time we need a new asset id, we just read that counter and then increment it.